DietPi-Software | CUPS

Joulinar opened the pull request:
CUPS new software title (ID 187)

**Status**: Ready
- [x] add new software title CUPS
- [x] add new service CUPS
- [x] add change log
- [x] add `avahi-daemon` support

**Testing**: install/reinstall/uninstall/printing
- [x] Buster RPi1 32bit
- [x] Buster RPi3B+ 32bit
- [x] Buster RPi4B 64bit
- [x] Bullseye VM
- [x] Buster VM
- [x] Stretch VM
- [x] test print on W10
- [x] test print on macOS
- [x] test print on Android
- [x] test print on iOS

**Reference**: https://github.com/MichaIng/DietPi/issues/3942

**Commit list/description**:
- DietPi-Software | CUPS: add new software title
- DietPi-Services | CUPS: add new software title
- DietPi-Software | CUPS: add `avahi-daemon` as dependency
- DietPi-Survey_report | Add CUPS support, newly added with DietPi v7.1
- DietPi-Software | CUPS - Download base configuration if it does not exist yet

Joulinar committed:
v7.1

DietPi-Software | CUPS: add new software title

Joulinar committed:
v7.1

DietPi-Services | CUPS: add new software title

Joulinar commented:
ok first attempt, with quite a simple installation procedure.

Joulinar committed:
v7.1

DietPi-Software | CUPS: add avahi-daemon as dependency

Joulinar commented:
at the moment I'm puzzled, struggling with printing on Android nativ printing functionality. There is no issue to detect printers on W10 as well as iOS. However Android is strange. I have 3 devices, a S8, S9 and a Pixel 4. While the S9 is working fine, printers are not detected on the S8 + Pixel 4 🤔

MichaIng commented:
Good that everything else works already. I'll test it on my Samsung A3 with Android 8.0 and two printers around here.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1880,6 +1880,14 @@ _EOF_
+		software_id=187
+
+		aSOFTWARE_NAME[$software_id]='CUPS'
+		aSOFTWARE_DESC[$software_id]='common UNIX Pprinting system'
```
Typo: `Pprinting` => `printing`

Joulinar replied:
will be updated once we agreed on installation procedure

MichaIng committed:
v7.1

+ DietPi-Survey_report | Add CUPS support, newly added with DietPi v7.1

MichaIng committed:
Merge branch 'dev' into cups

MichaIng committed:
v7.1

+ DietPi-Survey_report | Satisfy shellcheck

Joulinar commented:
@MichaIng 
One challenge, on reinstall we will apply local access configuration again. Is this ok or should we skip this step on reinstall?

MichaIng commented:
IMO we should leave custom configs untouched. The the package ships with a default config file, either we could set a variable prior to package install or check for some content that only exists on a new config file.

Joulinar commented:
by default LISTEN port is ~~0.0.0.0~~ `localhost`. We could assume that this is valid for new/untouched configs. Because usually people will change this to be able to access the CUPS 😉

MichaIng commented:
I see `Listen localhost:631`, so you mean `127.0.0.1`, I guess?

Joulinar commented:
ah sorry, my fault 😃 Yes it is `localhost` by default

MichaIng commented:
Not 100% failsafe, since it could be setup with reverse proxy like this, I guess?

Joulinar commented:
let me check the behaviour if config file exist before apt package installation

Joulinar commented:
ok an existing config will not be overwritten by the apt package installation. Means we could do it as follow

1. on an empty system check for availability of `cupsd.conf` 
2. if not present, pull our own `cupsd.conf` config file 
3. perform `apt install cups`

on reinstall
1. system check for availability of `cupsd.conf` 
2. if present, do not pull `cupsd.conf` config file
3. perform `apt reinstall cups`

Joulinar committed:
v7.1

DietPi-Software | CUPS - Download base configuration if it does not exist yet

MichaIng commented:
🈯 Works pretty fine on Buster VM, including web UI, adding my two test printers (USB pass-through), finding them on Windows 10 as network printers, printing test pages from Windows 10 and from web UI.

Now testing on Android 8, using the FOSS CUPS print Android app...
🈯 ... both printers are found as network printers.
🈯 ... I can't test actual printing Android as I freezed the print spooler app and I cannot unfreeze it without unfreezing all other bloat, as the related app got broken with last Android update 😄.

Testing on macOS...
🈯 ... both printers were found as network printers.
🈯 ... printing works fine 😃.
___________
🈯 Same on Debian Bullseye.
___________
Tested web access and permissions:
- Login works via every UNIX user
- Printing via web UI is only possible for members of `lp` group, e.g.: `usermod -aG lp dietpi`
- Adding/managing printers via web UI is only possible for members of `lpadmin` group, e.g.: `usermod -aG lpadmin dietpi`
- When managing printers, the access is automatically upgraded to HTTPS. At least with current Chromium-based browsers, this is not a problem since the missing trusted certificate warning can be ignored one time and that decision is remembered.
- So all works as expected regarding access permissions.
___________
🈯 Same on Debian Bullseye.
- The web UI does not show the login prompt immediately, but only when trying to add/manage printers. But that's fine, IMO.

MichaIng committed:
v7.1

+ CHANGELOG | CUPS: The common UNIX printing system is now available as optimised install option

Joulinar commented:
1. yes HTTPS is forced by CUPS if you like to access the admin area.
2. I will add `dietpi` group to user `lp` and `lpadmin`
3. yes login is needed on admin page only if you like to configure stuff. This is due to setting `Require user @SYSTEM` on the config file for these areas. If needed we could add this to the initial page as well

nonnorm commented:
I can verify that installing and adding a printer works on RPi 4B 64-bit.  I'll test printing with iOS later when I get home.

MichaIng commented:
> I will add `dietpi` group to user `lp` and `lpadmin`

You mean add `dietpi` user to `lp` and `lpadmin` groups? No please skip that. I don't want to add any further steps to make `dietpi` a required login user, instead I want to remove it and allow creating a custom user on first boot instead. Having that user where not all "users" know about, which can be used to login via SSH etc is a security issue. It's fine for now that OOTB root login works, and if users use any other login user, reasonably, then they need to add it to `lpadmin`, which is fine when documented.

An alternative I could think if is to create a new `cups` user that is explicitly created for CUPS web UI login, and CUPS could be configured to allow only that user and no other. But since it needs to be a login user with password, it can be used to login via SSH as well, so it would need to be isolated, disabling the login shell and use a `/nonexistent` home directory to not cause another attack vector 😉.

> yes login is needed on admin page only if you like to configure stuff. This is due to setting `Require user @SYSTEM` on the config file for these areas. If needed we could add this to the initial page as well

Ah, probably I did remember it from Buster and Bullseye tests wrong. It's fine IMO.

Joulinar commented:
ok in this case we need to finalize testing only and maybe check out Android

MichaIng commented:
> and maybe check out Android

As said, while detecting the printers works pretty fine, I broke the printing spooler app on my phone, so cannot test to send and actual print. Ah maybe I find time to test it with the phone of my wife later 😄.

Joulinar commented:
Question is if we should rely on Andrew native printing feature or if we can recomend an Android app for users (online docs) 🤔 Maybe I'm thinking to complicate atm

MichaIng commented:
Usually the native printing provider should work. But if users face issues, or have a custom OS with no printing service pre-installed, we could suggest to try:
- https://f-droid.org/packages/io.github.benoitduffez.cupsprint/
- https://play.google.com/store/apps/details?id=io.github.benoitduffez.cupsprint

Worked fine for me, although there are alternatives as well which I did not test.

nonnorm commented:
It worked for me, but per [this](https://askubuntu.com/questions/82016/network-printer-found-but-cannot-print#answer-455285) askubuntu answer I had to install `libnss-mdns` and change the lines in the config file.  This seems to only be required for Brother printers, but should we install it anyway?
**EDIT**: Notably, this let me `ssh` into my RPi using its `rpi4.local` address, which didn't work before, and this stayed after removing the package, and removing the entry from my `known_hosts` file.
**EDIT2**: Nope, turns out that this it's the Avahi Daemon that let me do this, not the `libnss-mdns` package.

Joulinar commented:
ok finished testing. However printing on Android is still a mystery for me. Maybe it's me or my devices, I don't know 🤣 It seems to be better on Bullseye. Anyway, ready from my side.

MichaIng commented:
> EDIT2: Nope, turns out that this it's the Avahi Daemon that let me do this, not the libnss-mdns package.

Good, I was just wondering since one of my successfully tested printers is a brother as well, though an old one 😄. Great I'll merge it then.

Joulinar commented:
I guess the 2nd update was more related to the fact being able to access the system using `rpi4.local`. the issue with Brother printers seems to be fixed by installing `libnss-mdns` if I'm not mistaken. But this I'm not able to test as I have a HP printer only.

MichaIng commented:
Hmm, but that indeed should not be related to CUPS at all, as it's only a question whether the hostname.local is made available in the network or not, so indeed the Avahi daemons task, which is pulled as dependency by [libnss-mdns](https://packages.debian.org/buster/libnss-mdns) but installed by `dietpi-software` already.

nonnorm commented:
Yes, you need `libnss-mdns` for Brother printers to work.  The hostname.local is completely unrelated.

MichaIng commented:
Strange, in my case if was not required, and I tested it with many different systems, client and host.

... ah I think now I understand in which case it might be required:
- When you add a network printer to CUPS that broadcasts it's own hostname via mDNS (like Avahi daemon), and you want to add it by this mDNS name. @ravenclaw900 does this match your case?
- I had all tested printers locally attached to the host system, respectively passed through to the VM.

I think it's fine to not install it by default but leave a hint in the docs for now. Usually, when you use CUPS, it is not advised to have printers additionally broadcast their own hostname via mDNS, else you would see the same printer two times in the network, one time broadcasted by itself, one time broadcasted by CUPS, isn't it? Probably it's mDNS feature can be disabled and it added via its IP address to CUPS instead to avoid such overhead/collisions.

nonnorm commented:
Yes, that's what I was testing.  For some reason, that printer wasn't actually showing by default on iOS, and CUPS was needed to make it appear.  However, when I tested with a printer that did appear, it wouldn't produce a duplicate entry.

Joulinar commented:
I guess there are quite some dependency on local setup and features the printer is providing. If you have a network printer reachable by it's own, do you realy need cups? 😉 

Let's see what people will use it for. I guess it's ok for first attempt and we can adjusted later on.
